### PR TITLE
add healthcheck router

### PR DESCRIPTION
**Pull request by CodingSinger**

### CODE REVIEW

**Comment by CodingSinger** [comment]
refer #337

File: cluster/router/condition/default_health_check.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
+package condition
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]: apache license

File: cluster/router/condition/default_health_check_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package condition
```
Comment (zouyx) [inline cluster/router/condition/default_health_check_test.go]: apache license

File: cluster/router/condition/health_check_route.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
+package condition
```
Comment (zouyx) [inline cluster/router/condition/health_check_route.go]: apache license

File: cluster/router/condition/health_check_route_test.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
+package condition
```
Comment (zouyx) [inline cluster/router/condition/health_check_route_test.go]: apache license

File: cluster/router/condition/default_health_check.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// isCircuitBreakerTripped determine whether the invoker is in the tripped state by the number of successive bad request
+func (c *DefaultHealthChecker) isCircuitBreakerTripped(status *protocol.RPCStatus) bool {
+	circuitBreakerTimeout := c.getCircuitBreakerTimeout(status)
+	currentTime := protocol.CurrentTimeMillis()
+	if circuitBreakerTimeout <= 0 {
+		return false
+	}
+	return circuitBreakerTimeout > currentTime
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]:
```suggestion
	return circuitBreakerTimeout > protocol.CurrentTimeMillis()
```

File: cluster/router/condition/default_health_check.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		return 0
+	}
+	return status.GetLastRequestFailedTimestamp() + sleepWindow
+}
+
+// getCircuitBreakerSleepWindowTime get the sleep window time of invoker
+func (c *DefaultHealthChecker) getCircuitBreakerSleepWindowTime(status *protocol.RPCStatus) int64 {
+
+	successiveFailureCount := status.GetSuccessiveRequestFailureCount()
+	diff := successiveFailureCount - c.RequestSuccessiveFailureThreshold
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]: use getter for get ```c.RequestSuccessiveFailureThreshold```

File: cluster/router/condition/default_health_check.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+func (c *DefaultHealthChecker) getCircuitBreakerTimeout(status *protocol.RPCStatus) int64 {
+	sleepWindow := c.getCircuitBreakerSleepWindowTime(status)
+	if sleepWindow <= 0 {
+		return 0
+	}
+	return status.GetLastRequestFailedTimestamp() + sleepWindow
+}
+
+// getCircuitBreakerSleepWindowTime get the sleep window time of invoker
+func (c *DefaultHealthChecker) getCircuitBreakerSleepWindowTime(status *protocol.RPCStatus) int64 {
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]: as above

File: cluster/router/condition/default_health_check.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	circuitBreakerTimeout := c.getCircuitBreakerTimeout(status)
+	currentTime := protocol.CurrentTimeMillis()
+	if circuitBreakerTimeout <= 0 {
+		return false
+	}
+	return circuitBreakerTimeout > currentTime
+}
+
+// getCircuitBreakerTimeout get the timestamp recovered from tripped state
+func (c *DefaultHealthChecker) getCircuitBreakerTimeout(status *protocol.RPCStatus) int64 {
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]: what's the unit for this time? specify in comment.

File: cluster/router/condition/default_health_check.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+const (
+	HEALTH_CHECKER                             = "health.checker"
+	DEFAULT_HEALTH_CHECKER                     = "default"
+	OUTSTANDING_REQUEST_COUNT_LIMIT_KEY        = "outstanding.request.limit"
+	SUCCESSIVE_FAILED_REQUEST_THRESHOLD_KEY    = "successive.failed.threshold"
+	DEFAULT_SUCCESSIVE_FAILED_THRESHOLD        = 5
+	CIRCUIT_TRIPPED_TIMEOUT_FACTOR_KEY         = "circuit.tripped.timeout.factor"
+	DEFAULT_SUCCESSIVE_FAILED_REQUEST_MAX_DIFF = 5
+	DEFAULT_CIRCUIT_TRIPPED_TIMEOUT_FACTOR     = 1000
+	MAX_CIRCUIT_TRIPPED_TIMEOUT                = 30000
```
Comment (zouyx) [inline cluster/router/condition/default_health_check.go]: if this variable is time , you can use  time.Duration ,like  ```time.Duration(sec) * time.Second```?

File: common/constant/key.go [hunk 10]
```diff
@@ -158,7 +158,8 @@ const (
 	// ConditionAppRouterName Specify listenable application router name
 	ConditionAppRouterName = "app"
 	// ListenableRouterName Specify listenable router name
-	ListenableRouterName = "listenable"
+	ListenableRouterName  = "listenable"
+	HealthCheckRouterName = "health_check"
```
Comment (zouyx) [inline common/constant/key.go]: add comment for it

File: cluster/router/condition/default_health_check.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	"github.com/apache/dubbo-go/cluster/router"
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/protocol"
+)
+
+const (
+	HEALTH_CHECKER                             = "health.checker"
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: These variables may be used by the user.

File: cluster/router/condition/default_health_check.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+func init() {
+	extension.SethealthChecker(DEFAULT_HEALTH_CHECKER, NewDefaultHealthChecker)
+}
+
+// DefaultHealthChecker is the default implementation of HealthChecker, which determines the health status of
+// the invoker based on the number of successive bad request and the current active request.
+type DefaultHealthChecker struct {
+	// OutStandingRequestConutLimit
+	OutStandingRequestConutLimit int32
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: Thanks for your suggestion.

File: cluster/router/condition/default_health_check.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	RequestSuccessiveFailureThreshold int32
+	// RequestSuccessiveFailureThreshold
+	CircuitTrippedTimeoutFactor int32
+}
+
+// IsHealthy evaluates the healthy state on the given Invoker based on the number of successive bad request
+// and the current active request
+func (c *DefaultHealthChecker) IsHealthy(invoker protocol.Invoker) bool {
+	urlStatus := protocol.GetURLStatus(invoker.GetUrl())
+	if c.isCircuitBreakerTripped(urlStatus) || urlStatus.GetActive() > c.OutStandingRequestConutLimit {
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: Why not just use `c. outStandingRequestConutLimit ` to get the value?

File: cluster/router/condition/default_health_check.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router/condition/default_health_check.go]: i think you should put those variable to ``` constant ``` package, if not will easy cause circular reference

File: cluster/router/condition/default_health_check.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router/condition/default_health_check.go]: ``` getter ``` provide a unified export, you can easy to refactor it.

File: cluster/router/condition/default_health_check.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: done

File: cluster/router/condition/default_health_check.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: done

File: cluster/router/condition/default_health_check.go (hunk 6 above)
Comment (CodingSinger) [reply cluster/router/condition/default_health_check.go]: done

File: cluster/router/healthcheck/health_check_route.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+	url     *common.URL
+	enabled bool
+	checker router.HealthChecker
+}
+
+// NewHealthCheckRouter construct an HealthCheckRouter via url
+func NewHealthCheckRouter(url *common.URL) (router.Router, error) {
+	r := &HealthCheckRouter{}
+	r.url = url
+	r.enabled = url.GetParamBool(HEALTH_ROUTE_ENABLED_KEY, false)
```
Comment (AlexStocks) [inline cluster/router/healthcheck/health_check_route.go]: u can init url and enabled in this way

r := &HealthCheckRouter{
	 url : url,
	 enabled : url.GetParamBool(HEALTH_ROUTE_ENABLED_KEY, false),
     }

File: cluster/router/healthcheck/health_check_route.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	return r, nil
+}
+
+// Route gets a list of healthy invoker
+func (r *HealthCheckRouter) Route(invokers []protocol.Invoker, url *common.URL, invocation protocol.Invocation) []protocol.Invoker {
+	if !r.enabled {
+		return invokers
+	}
+	var healthyInvokers []protocol.Invoker
```
Comment (AlexStocks) [inline cluster/router/healthcheck/health_check_route.go]: ealthyInvokers = make([]protocol.Invoker, 0, len(invokers))

File: cluster/router/healthcheck/health_check_route.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	for _, invoker := range invokers {
+		if r.checker.IsHealthy(invoker) {
+			healthyInvokers = append(healthyInvokers, invoker)
+		}
+	}
+	// If all Invoke are considered unhealthy, downgrade to all inovker
+	if len(healthyInvokers) == 0 {
+		logger.Warnf(" Now all invokers are unhealthy, so downgraded to all! Service: [%s]", url.ServiceKey())
+		return invokers
+	} else {
```
Comment (AlexStocks) [inline cluster/router/healthcheck/health_check_route.go]: delete the else branch

File: common/constant/key.go [hunk 17]
```diff
@@ -188,3 +190,16 @@ const (
 	ACCESS_KEY_ID_KEY             = "accessKeyId"
 	SECRET_ACCESS_KEY_KEY         = "secretAccessKey"
 )
+
+// HealthCheck Router
+const (
```
Comment (AlexStocks) [inline common/constant/key.go]: pls add comments for the following constants.

File: cluster/router/healthcheck/default_health_check.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+// DefaultHealthChecker is the default implementation of HealthChecker, which determines the health status of
+// the invoker based on the number of successive bad request and the current active request.
+type DefaultHealthChecker struct {
+	// outStandingRequestConutLimit
+	outStandingRequestConutLimit int32
+	// requestSuccessiveFailureThreshold
+	requestSuccessiveFailureThreshold int32
+	// requestSuccessiveFailureThreshold
```
Comment (zouyx) [inline cluster/router/healthcheck/default_health_check.go]: // circuitTrippedTimeoutFactor

File: cluster/router/healthcheck/health_check_route.go (hunk 14 above)
Comment (CodingSinger) [reply cluster/router/healthcheck/health_check_route.go]: done.

File: cluster/router/healthcheck/health_check_route.go (hunk 16 above)
Comment (CodingSinger) [reply cluster/router/healthcheck/health_check_route.go]: good suggestion.

**Comment by AlexStocks** [comment]
LGTM

**Pull request merged**